### PR TITLE
subtracting numbers

### LINKED ISSUE #22: subtracting numbers

### PR DESCRIPTION
**Pull request by HafizHussein**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [ ] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_subtraction.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import unittest
+
+from solutions.subtraction import subtract
+
```
Comment (sky0-0glitch) [inline solutions/tests/test_subtraction.py:6]: Add a module-level docstring at the beginning of the test file to describe its purpose.

File: solutions/tests/test_subtraction.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        actual = subtract(10.5, 2.5, 3.0)
+        expected = 5.0
+        self.assertEqual(actual, expected)
+
+    """It should return the subtraction of zero numbers"""
+
+    def test_subtract_zero_numbers(self):
+        actual = subtract(0, 0, 0)
+        expected = 0
+        self.assertEqual(actual, expected)
```
Comment (sky0-0glitch) [inline solutions/tests/test_subtraction.py:42]: Add tests to check for invalid inputs and raise appropriate errors

Comment (sky0-0glitch) [inline solutions/tests/test_subtraction.py:42]: Include boundary case tests, such as no arguments and single arguments

File: solutions/subtraction.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+"""
+
+
+def subtract(*args):
+    """
+    Calculate the result of subtracting a series of numbers.
+
+
+    Parameters:
+        *args (int or float): Numbers to be subtracted in sequence.
```
Comment (sky0-0glitch) [inline solutions/subtraction.py:27]: update the docstring to address assumptions about *args something like:
Parameters:
    *args (int or float): Numbers to be subtracted in sequence. All inputs must be numeric (int or float).
Raises:
    TypeError: If any input is not a number.

File: solutions/subtraction.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    Returns:
+    int or float: The result obtained after performing the subtraction.
+
+
+    Example:
+    >>> subtract(10, 2, 3)
+    5
+
```
Comment (sky0-0glitch) [inline solutions/subtraction.py:42]: Add more doctests to the docstring you should have three or more.

File: solutions/subtraction.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+@author: Hafiz Hussein
+
+Inspired by a challenge from the Leetcode website.
+
+Explanation of subtraction for a given number:
+
+"""
+
+
+def subtract(*args):
```
Comment (sky0-0glitch) [inline solutions/subtraction.py]: Add proper type annotations to the function.

Comment (sky0-0glitch) [inline solutions/subtraction.py]: Add a brief comment in the function explaining the approach and strategy

Comment (sky0-0glitch) [inline solutions/subtraction.py]: you are missing defensive assertions you should add checks to ensure all inputs are numbers

**Review by sky0-0glitch — CHANGES_REQUESTED** [review]

File: solutions/tests/test_subtraction.py (hunk 1 above)
Comment (HafizHussein) [reply solutions/tests/test_subtraction.py:6]: created

File: solutions/subtraction.py (hunk 4 above)
Comment (HafizHussein) [reply solutions/subtraction.py:42]: added.

File: solutions/subtraction.py (hunk 5 above)
Comment (HafizHussein) [reply solutions/subtraction.py]: Added.

Comment (HafizHussein) [reply solutions/subtraction.py]: Added checks.

Comment (HafizHussein) [reply solutions/subtraction.py]: Commented.

File: solutions/subtraction.py (hunk 3 above)
Comment (HafizHussein) [reply solutions/subtraction.py:27]: updated it.

File: solutions/tests/test_subtraction.py (hunk 2 above)
Comment (HafizHussein) [reply solutions/tests/test_subtraction.py:42]: tests added.

Comment (HafizHussein) [reply solutions/tests/test_subtraction.py:42]: included.

File: solutions/subtraction.py (hunk 5 above)
Comment (sky0-0glitch) [reply solutions/subtraction.py]: Dear Hafiz you should write something like     def subtract(*args: int) -> int: 
here is a  little explanation from [runestone.academy](https://runestone.academy/ns/books/published/fopp/Functions/TypeAnnotations.html) to help you understand the concept better " A type annotation, sometimes called a type hint, is an optional notation that specifies the type of a parameter or function result. It tells the programmer using the function what kind of data to pass to the function, and what kind of data to expect when the function returns a value."

Comment (HafizHussein) [reply solutions/subtraction.py]: Thank you very much for your clarification.
I added the annotation.

Comment (sky0-0glitch) [reply solutions/subtraction.py]: great job 🙌

**Review by sky0-0glitch — APPROVED** [review]
